### PR TITLE
Migrate PARL codes and add examples of how to use Manager to run.

### PR DESCRIPTION
**Pull request by skylian**
the folder "parl" is temporarily moved here for our convenience.

### CODE REVIEW

File: parl/agent_zoo/simple_rl_agents.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    
+    By using different AgentHelpers, this Agent can be applied to either on-
+    policy or off-policy RL algorithms.
+    """
+
+    def __init__(self, env, num_games):
+        super(SimpleRLAgent, self).__init__(env, num_games)
+        self.log_q = None
+
+    @classmethod
```
Comment (yu239) [inline parl/agent_zoo/simple_rl_agents.py]: Should better make a general pack_exps and unpack_exp_seqs

File: parl/agent_zoo/simple_rl_agents.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+                episode_end=[episode_end])
+            obs = next_obs
+            episode_end = next_episode_end
+            if episode_end:
+                break
+        # we put dummy action and reward here as they will not be used
+        self.store_data(
+            'RL',
+            sensor=obs,
+            action=[0],
```
Comment (yu239) [inline parl/agent_zoo/simple_rl_agents.py]: Still need to compute the action for the last time step, although the action will not be used to step the env.

File: parl/algorithm_zoo/simple_algorithms.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/algorithm_zoo/simple_algorithms.py]: remove the file

File: parl/common/data_process.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/common/data_process.py]: remove the file

File: parl/common/tests/test_data_process.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/common/tests/test_data_process.py]: remove the file

File: parl/examples/ac_example.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/examples/ac_example.py]: remove the file

File: parl/framework/algorithm.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/framework/algorithm.py]: remove the file

File: parl/framework/computation_task.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/framework/computation_task.py]: remove the file

File: parl/framework/computation_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        with self.lock:
+            ret = self.ct.predict(**data)
+        ret = self._unpack_data(ret, starts)
+        return ret
+
+    def do_one_training(self, data):
+        data, starts = self._pack_data(data)
+        with self.lock:
+            ret = self.ct.learn(**data)
+        #ret = self._unpack_data(ret, starts)
```
Comment (yu239) [inline parl/framework/computation_wrapper.py]: uncomment

File: parl/framework/manager.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/framework/manager.py]: remove the file

File: parl/framework/policy_distribution.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/framework/policy_distribution.py]: remove file

File: parl/framework/tests/test_algorithm.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/framework/tests/test_algorithm.py]: remove file

File: parl/framework/tests/test_computation_task.py [hunk 13]
```diff
@@ -0,0 +1,245 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/framework/tests/test_computation_task.py]: remove file

File: parl/framework/tests/test_simple_games.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/framework/tests/test_simple_games.py]: remove file

File: parl/layers/__init__.py [hunk 15]
```diff
@@ -0,0 +1,19 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/layers/__init__.py]: remove the file

File: parl/layers/common_functions.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/layers/common_functions.py]: remove file

File: parl/layers/layer_wrappers.py [hunk 17]
```diff
@@ -0,0 +1,563 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/layers/layer_wrappers.py]: remove layers

File: parl/model_zoo/__init__.py [hunk 18]
```diff
@@ -0,0 +1,13 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yu239) [inline parl/model_zoo/__init__.py]: remove model_zoo

File: flare/examples/ac_example.py [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+    # 3. Specify the settings for learning: the algorithm to use (SimpleAC 
+    # in this case), data sampling strategy (OnPolicyHelper here) and other 
+    # settings used by ComputationTask.
+    ct_settings = {
+        "RL": dict(
+            algorithm=alg,
+            hyperparas=dict(lr=1e-4),
+            # sampling
+            sample_method=OnPolicyHelper,
+            sample_interval=8,
```
[the remark this answers is not in the document]
Comment (skylian) [reply flare/examples/ac_example.py]: there is no "batch_size" for on-policy algorithms

File: flare/examples/ac_example.py [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+            # ct wrapper
+            min_batchsize=1,
+            max_batchsize=16,
+            timeout=1)
```
[the remark this answers is not in the document]
Comment (skylian) [reply flare/examples/ac_example.py]: renamed to comm_timeout

File: flare/examples/q_example.py [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+    num_games = 8000
+    # 1. Create environments
+    envs = []
+    for _ in range(num_agent):
+        envs.append(gym.make(game))
+    state_shape = envs[-1].observation_space.shape[0]
+    num_actions = envs[-1].action_space.n
+
+    # 2. Construct the network and specify the algorithm.
+    #    Here we use a small MLP and apply the Actor-Critic algorithm
```
[the remark this answers is not in the document]
Comment (skylian) [reply flare/examples/q_example.py]: done

File: flare/examples/q_example.py [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+    # settings used by ComputationTask.
+    ct_settings = {
+        "RL": dict(
+            algorithm=alg,
+            hyperparas=dict(lr=1e-4),
+            # sampling
+            sample_method=ExpReplayHelper,
+            buffer_capacity=100000,
+            batch_size=32,
+            num_seqs=0,  # sample instances
```
[the remark this answers is not in the document]
Comment (skylian) [reply flare/examples/q_example.py]: `num_seqs=0` means instance sampling

File: flare/framework/computation_task.py [hunk 23]
```diff
@@ -4,6 +4,7 @@
 import recurrent as rc
 import numpy as np
 import operator
+from parl.framework.computation_wrapper import ComputationWrapper
```
[the remark this answers is not in the document]
Comment (skylian) [reply flare/framework/computation_task.py]: done

File: flare/framework/manager.py [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+from flare.framework.computation_task import ComputationTask
+from parl.common.logging import GameLogger
+
+
+class Manager(object):
+    def __init__(self, ct_settings):
+        self.agents = []
+        self.cts = {}
+        self.state_specs = {}
+        self.wrappers = {}
```
[the remark this answers is not in the document]
Comment (skylian) [reply flare/framework/manager.py]: renamed to `CDPs`, i.e., `ComputationDataProcessor`

File: flare/framework/manager.py [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+# limitations under the License.
+
+from multiprocessing import Queue
+from threading import Thread
+from flare.framework.computation_task import ComputationTask
+from parl.common.logging import GameLogger
+
+
+class Manager(object):
+    def __init__(self, ct_settings):
```
[the remark this answers is not in the document]
Comment (skylian) [reply flare/framework/manager.py]: done

File: flare/framework/manager.py [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+        self.logger = GameLogger(1, 100)
+
+    def add_agent(self, agent):
+        agent.id = len(self.agents)
+        # `Agent` needs to know the state specs to prepare state data
+        agent.make_initial_states(self.state_specs)
+        self.agents.append(agent)
+        for name, wrapper in self.wrappers.iteritems():
+            agent.add_helper(
+                wrapper.create_helper(agent.id), agent.pack_exps,
```
[the remark this answers is not in the document]
Comment (skylian) [reply flare/framework/manager.py]: done

File: parl/agent_zoo/simple_rl_agents.py (hunk 2 above)
Comment (skylian) [reply parl/agent_zoo/simple_rl_agents.py]: done

File: flare/framework/manager.py [hunk 27]
```diff
@@ -20,24 +20,29 @@
...
     def add_agent(self, agent):
         agent.id = len(self.agents)
         # `Agent` needs to know the state specs to prepare state data
         agent.make_initial_states(self.state_specs)
         self.agents.append(agent)
-        for name, wrapper in self.wrappers.iteritems():
-            agent.add_helper(
-                wrapper.create_helper(agent.id), agent.pack_exps,
+        for name, cdp in self.CDPs.iteritems():
+            agent.add_agent_helper(
```
Comment (yu239) [inline flare/framework/manager.py]: My original comment was meant to change cdp.create_helper to cdp.create_agent_helper, not agent.add_helper

File: parl/framework/agent.py [hunk 28]
```diff
@@ -16,16 +16,14 @@
...
 from parl.common.replay_buffer import NoReplacementQueue, ReplayBuffer
 
 
 class AgentHelper(object):
     """
     AgentHelper abstracts some part of Agent's data processing and the I/O 
-    communication between Agent and ComputationWrapper. It receives a
-    Communicator from one ComputationWrapper and uses it to send data to the
-    ComputationWrapper.
+    communication between Agent and ComputationDataProcessor (CP). It receives a
```
Comment (yu239) [inline parl/framework/agent.py]: CP -> CDP

File: parl/framework/computation_data_processor.py [hunk 29]
```diff
@@ -19,23 +19,23 @@
...
-class ComputationWrapper(object):
+class ComputationDataProcessor(object):
     """
     This class batches the data from the Agent side (i.e., sent by AgentHelper)
     and sends them to ComputationTask. At the time of results returned, it
     splits the batched results and sends them back to AgentHelper.
     """
 
-    def __init__(self, name, ct, min_batchsize, max_batchsize, timeout,
+    def __init__(self, name, ct, min_batchsize, max_batchsize, comm_timeout,
```
Comment (yu239) [inline parl/framework/computation_data_processor.py]: min_batchsize and max_batchsize should be both num_agents;

File: flare/examples/rnn_ac_example.py [hunk 30]
```diff
@@ -57,8 +57,7 @@
             sample_interval=8,
             # ct wrapper
             min_batchsize=1,
-            max_batchsize=16,
-            timeout=1)
+            max_batchsize=16)
```
Comment (yu239) [inline flare/examples/rnn_ac_example.py]: These two arguments should be removed?

File: parl/framework/agent.py [hunk 31]
```diff
@@ -122,12 +116,12 @@ class ExpReplayHelper(AgentHelper):
     run learn().
     """
 
-    def __init__(self, name, communicator, buffer_capacity, batch_size,
+    def __init__(self, name, communicator, buffer_capacity, sample_size,
```
Comment (yu239) [inline parl/framework/agent.py]: sample_size -> samples_n

File: parl/framework/computation_data_processor.py [hunk 32]
```diff
@@ -123,16 +118,19 @@ def _prediction_loop(self):
...
+        agent_ids = []
+        data = []
         while not self.exit_flag:
-            agent_ids = []
-            data = []
             try:
-                while (len(agent_ids) < self.min_batchsize or
-                       len(agent_ids) < self.max_batchsize and
+                while (len(agent_ids) < self.min_agents_per_batch or
+                       len(agent_ids) >= self.min_agents_per_batch and
```
Comment (yu239) [inline parl/framework/computation_data_processor.py]: should better use "(" to group the condition

**Pull request merged**